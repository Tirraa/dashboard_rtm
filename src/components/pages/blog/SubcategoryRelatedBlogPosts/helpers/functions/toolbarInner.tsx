import type { MaybeNull } from '@rtm/shared-types/CustomUtilityTypes';
import type { Quantity, Count, Id } from '@rtm/shared-types/Numbers';
import type { BlogTag } from '##/config/contentlayer/blog/blogTags';
import type { MutableRefObject, ReactElement } from 'react';
import type { FiltersAssoc } from '@/config/Blog/client';
import type { AppPath } from '@rtm/shared-types/Next';
import type { BlogTagId } from '@/types/Blog';

import { PAGE_KEY } from '@/components/ui/helpers/PaginatedElements/constants';

import type { ETagsSwitch } from '../enums';

import { buildDropdown } from '../../../helpers/functions/paginationWidget';
import shouldShowTagsCommandWidget from './shouldShowTagsCommandWidget';
import shouldShowPaginationWidget from './shouldShowPaginationWidget';
import FiltersSelectWidget from '../../../FiltersSelectWidget';
import TagsCommandWidget from '../../../TagsCommandWidget';
import PaginationWidget from '../../../PaginationWidget';

// eslint-disable-next-line no-magic-numbers
const shouldShowFiltersSelectWidget = (postsAmount: Quantity): boolean => postsAmount > 1;

const getMaybeDropdown = (
  showPaginationWidget: boolean,
  pagesAmount: Quantity,
  currentPage: Count,
  pathname: AppPath,
  searchParams: URLSearchParams,
  isBottomWidget?: boolean
): MaybeNull<ReactElement> =>
  !showPaginationWidget
    ? null
    : buildDropdown(pagesAmount, currentPage, pathname, searchParams, PAGE_KEY, { dropdownContentClassName: 'z-20', isBottomWidget });

function buildWidgetsForTopRight(
  paginationWidgetProps: {
    showPaginationWidget: boolean;
    pagesAmount: Quantity;
    currentPage: Count;
  },
  filtersWidgetProps: {
    newSelectedFilter: MutableRefObject<MaybeNull<Id>>;
    setSelectedFilterSwitch: (s: boolean) => unknown;
    showFiltersSelectWidget: boolean;
    filtersAssoc: FiltersAssoc;
    selectedFilter: Id;
  }
): ReactElement[] {
  const elements: ReactElement[] = [];

  if (filtersWidgetProps.showFiltersSelectWidget) {
    elements.push(
      <FiltersSelectWidget
        setSelectedFilterSwitch={filtersWidgetProps.setSelectedFilterSwitch}
        newSelectedFilter={filtersWidgetProps.newSelectedFilter}
        selectedFilter={filtersWidgetProps.selectedFilter}
        filtersAssoc={filtersWidgetProps.filtersAssoc}
        triggerClassName="z-20 mb-1 self-end"
        selectContentClassName="z-20"
        key="filters-widget"
      />
    );
  }

  if (paginationWidgetProps.showPaginationWidget) {
    elements.push(
      <PaginationWidget
        pagesAmount={paginationWidgetProps.pagesAmount}
        currentPage={paginationWidgetProps.currentPage}
        className="w-full justify-end"
        key="pagination-widget"
      />
    );
  }

  return elements;
}

function buildWidgetsForBottomRight(
  paginationWidgetProps: {
    showPaginationWidget: boolean;
    pagesAmount: Quantity;
    currentPage: Count;
  },
  pathname: AppPath,
  searchParams: URLSearchParams
): ReactElement[] {
  const elements: ReactElement[] = [];
  const isBottomWidget = true;

  const maybeDropdown = getMaybeDropdown(
    paginationWidgetProps.showPaginationWidget,
    paginationWidgetProps.pagesAmount,
    paginationWidgetProps.currentPage,
    pathname,
    searchParams,
    isBottomWidget
  );

  if (maybeDropdown !== null) elements.push(maybeDropdown);

  return elements;
}

export function buildTopLeftWidgets(tagsCommandWidgetProps: {
  memorizedPageBeforeFiltering: MutableRefObject<MaybeNull<Id>>;
  newSelectedTagsIds: MutableRefObject<MaybeNull<Id[]>>;
  setSelectedTagSwitch: (s: ETagsSwitch) => unknown;
  selectedTagsIds: BlogTagId[];
  tags: BlogTag[];
}): ReactElement[] {
  const showTagsCommandWidget = shouldShowTagsCommandWidget(tagsCommandWidgetProps.tags.length);

  const elements: ReactElement[] = [];

  if (showTagsCommandWidget) {
    elements.push(
      <TagsCommandWidget
        memorizedPageBeforeFiltering={tagsCommandWidgetProps.memorizedPageBeforeFiltering}
        setSelectedTagSwitch={tagsCommandWidgetProps.setSelectedTagSwitch}
        newSelectedTagsIds={tagsCommandWidgetProps.newSelectedTagsIds}
        selectedTagsIds={tagsCommandWidgetProps.selectedTagsIds}
        tags={tagsCommandWidgetProps.tags}
        key="tags-filter-widget"
      />
    );
  }

  return elements;
}

export function buildTopRightWidgets(
  paginationWidgetProps: {
    pagesAmount: Quantity;
    postsAmount: Quantity;
    currentPage: Count;
  },
  filtersWidgetProps: {
    newSelectedFilter: MutableRefObject<MaybeNull<Id>>;
    setSelectedFilterSwitch: (s: boolean) => unknown;
    filtersAssoc: FiltersAssoc;
    selectedFilter: Id;
  }
): ReactElement[] {
  const showPaginationWidget = shouldShowPaginationWidget(paginationWidgetProps.pagesAmount);
  const showFiltersSelectWidget = shouldShowFiltersSelectWidget(paginationWidgetProps.postsAmount);

  return buildWidgetsForTopRight(
    {
      pagesAmount: paginationWidgetProps.pagesAmount,
      currentPage: paginationWidgetProps.currentPage,
      showPaginationWidget
    },
    {
      setSelectedFilterSwitch: filtersWidgetProps.setSelectedFilterSwitch,
      newSelectedFilter: filtersWidgetProps.newSelectedFilter,
      selectedFilter: filtersWidgetProps.selectedFilter,
      filtersAssoc: filtersWidgetProps.filtersAssoc,
      showFiltersSelectWidget
    }
  );
}

export function buildBottomRightWidgets(
  pathname: AppPath,
  searchParams: URLSearchParams,
  paginationWidgetProps: {
    pagesAmount: Quantity;
    currentPage: Count;
  }
): ReactElement[] {
  const showPaginationWidget = shouldShowPaginationWidget(paginationWidgetProps.pagesAmount);

  return buildWidgetsForBottomRight(
    {
      pagesAmount: paginationWidgetProps.pagesAmount,
      currentPage: paginationWidgetProps.currentPage,
      showPaginationWidget
    },
    pathname,
    searchParams
  );
}
