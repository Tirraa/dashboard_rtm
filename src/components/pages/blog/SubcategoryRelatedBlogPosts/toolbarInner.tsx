'use client';

import type { FunctionComponent } from 'react';

import type { TagsFiltersWidgetProps } from '../TagsFiltersWidget';
import type { PaginationWidgetProps } from '../PaginationWidget';

import TagsFiltersWidget from '../TagsFiltersWidget';
import PaginationWidget from '../PaginationWidget';

export interface SubcategoryRelatedBlogPostsClientToolbarInnerProps extends TagsFiltersWidgetProps, PaginationWidgetProps {}

const SubcategoryRelatedBlogPostsClientToolbarInner: FunctionComponent<SubcategoryRelatedBlogPostsClientToolbarInnerProps> = ({
  setSelectedTagsIds,
  expectedTagsIds,
  selectedTagsIds,
  maxPagesAmount,
  pagesAmount,
  extraCtx,
  maxId,
  tags
}) => (
  <nav className="my-4 flex items-center justify-between">
    <TagsFiltersWidget
      setSelectedTagsIds={setSelectedTagsIds}
      selectedTagsIds={selectedTagsIds}
      expectedTagsIds={expectedTagsIds}
      maxPagesAmount={maxPagesAmount}
      extraCtx={extraCtx}
      maxId={maxId}
      tags={tags}
    />
    <PaginationWidget pagesAmount={pagesAmount} />
  </nav>
);

export default SubcategoryRelatedBlogPostsClientToolbarInner;
