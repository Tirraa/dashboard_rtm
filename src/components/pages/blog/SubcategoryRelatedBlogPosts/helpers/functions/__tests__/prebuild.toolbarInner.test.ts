import { describe, expect, it } from 'vitest';

import { buildBottomRightWidgets } from '../toolbarInner';

describe('buildBottomRightWidgets', () => {
  it('should return an empty list, given no reason to generate a pagination dropdown', () => {
    const pathname = 'pathname';
    const searchParams = new URLSearchParams();

    const pagesAmount = 1;
    const currentPage = 1;

    const bottomRightWidgets = buildBottomRightWidgets(pathname, searchParams, { pagesAmount, currentPage });
    expect(bottomRightWidgets).toStrictEqual([]);
  });

  it('should match snapshot', () => {
    const pathname = 'pathname';
    const searchParams = new URLSearchParams();

    const pagesAmount = 4;
    const currentPage = 1;

    const bottomRightWidgets = buildBottomRightWidgets(pathname, searchParams, { pagesAmount, currentPage });
    expect(bottomRightWidgets).toMatchSnapshot();
  });
});
