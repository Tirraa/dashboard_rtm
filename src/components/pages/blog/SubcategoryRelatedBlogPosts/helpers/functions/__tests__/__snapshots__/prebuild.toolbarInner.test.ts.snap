// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildBottomRightWidgets > should match snapshot 1`] = `
[
  <PaginationEllipsis
    dropdownContentClassName="z-20"
    dropdownItems={
      [
        <DropdownMenuItem
          className="h-10 p-0"
          onClick={[Function]}
        >
          <ForwardRef(LinkComponent)
            aria-current="page"
            className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold rounded-md bg-primary"
            href="pathname?"
            title="1"
          >
            1
          </ForwardRef(LinkComponent)>
        </DropdownMenuItem>,
        <DropdownMenuItem
          className="h-10 p-0"
          onClick={[Function]}
        >
          <ForwardRef(LinkComponent)
            className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold"
            href="pathname?page=2"
            title="2"
          >
            2
          </ForwardRef(LinkComponent)>
        </DropdownMenuItem>,
        <DropdownMenuItem
          className="h-10 p-0"
          onClick={[Function]}
        >
          <ForwardRef(LinkComponent)
            className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold"
            href="pathname?page=3"
            title="3"
          >
            3
          </ForwardRef(LinkComponent)>
        </DropdownMenuItem>,
        <DropdownMenuItem
          className="h-10 p-0"
          onClick={[Function]}
        >
          <ForwardRef(LinkComponent)
            className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold"
            href="pathname?page=4"
            title="4"
          >
            4
          </ForwardRef(LinkComponent)>
        </DropdownMenuItem>,
      ]
    }
    isBottomWidget={true}
    pageNumberIndicator={1}
  />,
]
`;

exports[`buildBottomRightWidgets should match snapshot 1`] = `
[
  <NoName key="ellipsis" dropdownContentClassName="z-20" pageNumberIndicator=1 isBottomWidget=true dropdownItems=[
      <NoName key="page-1" onClick=[Function] className="h-10 p-0">
        <NoName className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold rounded-md bg-primary" aria-current="page" title="1" href="pathname?" />
      </NoName>,
      <NoName key="page-2" onClick=[Function] className="h-10 p-0">
        <NoName className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold" aria-current=undefined title="2" href="pathname?page=2" />
      </NoName>,
      <NoName key="page-3" onClick=[Function] className="h-10 p-0">
        <NoName className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold" aria-current=undefined title="3" href="pathname?page=3" />
      </NoName>,
      <NoName key="page-4" onClick=[Function] className="h-10 p-0">
        <NoName className="flex h-full w-full items-center justify-center border-none px-2 text-center font-bold" aria-current=undefined title="4" href="pathname?page=4" />
      </NoName>,
    ] />,
]
`;
